### ISSUE TITLE
Should Registry() always be a singleton

### ISSUE THREAD

**Issue by dhobern** [aus-plant-phenomics-network/miappe_packaging#2]
At present, the implementation of appn.core.struct.Registry is always as a singleton. This works well in many contexts and makes it easy for the LinkedDataClass instances to self-register, but could presumably cause issues later.

Situations that might cause problems:

1. Packaging situations where we wish to generate two RO-Crates, one for the primary data and one for the derived assets, or full data versus primary research elements. It could be valuable to have two RO-Crates open simultaneously to allow them to be populated in parallel or to copy a subset of objects from one to another.
2. Long-running processors that perform tasks on a series of RO-Crates in order.
3. Multi-threaded use?

I don't yet know what the best trade-off would be. It could involve keeping the current implementation as a default, but with alternative processing at various points to allow multiple instances, e.g.:

- allowing explicit creation of non-singleton Registry objects (presumably with another parameter on the constructor)
- only using the singleton implementation if no non-singletons have been created
- requiring explicit calls to register each instance or use of an additional constructor parameter on LinkedDataClass instances for specifying the Registry

Either way a Registry.flush() method or similar could be useful to reset state for a new graph.

**Comment by harryle95** [comment]
Yes you are correct. Originally I planned to make Registry an all-in-one package for better tracking of URI references and validation but I am progressively finding it more and more restrictive. I am moving to having a 
- Registry object which manages LinkedDataClass - rdf resource mapping, and a 
- Session object which manages all the active instances to be serialised 
Registry will be used to parse raw rdf triples to LinkedDataClass objects when loading from a file or over a network, while Session will act like a db interface with commit, rollback, and flush.

**Comment by dhobern** [comment]
Probably just stating the obvious, but how do you then expect to manage the associate between LinkedDataClass objects and Session objects? If you shift away from automatic binding of all LinkedDataClass objects to a singleton and have a method or parameter for making the association, you'll also need to consider likely user behaviours. 

To keep it simple, this example will pretend we always use an explicit Registry.register() method to make the association.

I can imagine users trying the following:

`plant = Plant(variety="barley-alpha")`
`for i in range(1, 9):`
`    plant.id = f"Plant_{i}"`
`    registry.register(plant)`

All eight plants would currently be references to the same instance. Do we need to be more defensive and make a copy each time we register one?

**Comment by harryle95** [comment]
I am just looking through the sqlalchemy source code for implementation details, but I expect the Registry + Session model -i.e behaviour, usage, interfaces to follow theirs. 

For your example, plants 1 to 8 will be 8 separate instances and will be treated as 8 separate objects. 

I am writing this down as a design reference, some of which can also go into the documentation (coming soon), so please let me know if you have any feedback on either the design decision or the clarity of the documentation. 

## High-level overview

- `Registry` is a singleton that maps resource to LinkedDataClass. 
  - Whenever a LinkedDataClass subclass is created, the class and its schema are registered to the registry. 
- `Session` represents a connection to a data store, which can be a remote SPARQL endpoint, or in many cases an `RO-Crate` json file. 
  - `Session` wrap the `rdflib.Graph` object and provide features for 
    - querying and updating the graph using LinkedDataClass attributes - i.e. find all Plant instances whose variety is barley.
    - managing ID - i.e. check whether an internal reference is valid, bulk update ID based on some mapping rules
    - serialising and deserialising.

### Creating a LinkedDataClass subclass 

Users can create a subclass of `LinkedDataClass` to describe their experiment data models. By default, miappe classes will be provided, but can be subclassed to add extra fields. For instance, if uni A has an internal protocol ID, they can subclass `Investigation`: 

#### With explicit schema 

```python
class UniAInvestigation(Investigation):
    __schema__ = Schema(
        base=Investigation.__schema__,
        extra_attrs={"protocol_id": FieldInfo(ref=SchemaORG.protocol)},
    )
    protocol_id: str
```

Here, an explicit `__schema__` value is provided.

#### With implicit schema

```python
class UniAInvestigation(Investigation):
    protocol_id: Annotated[str, FieldInfo(ref=SchemaORG.protocol)]
```

Here, additional metadata is provided using `Annotated`.

#### With a factory method 

```python
UniAInvestigation = create_linked_dataclass(schema=schema, base=Investigation)
```
Here, schema can be a `Schema` object, or an external uri

#### LinkedDataClass behaviours

Whenever a subclass of `LinkedDataClass` is created, its `__rdf_resource__` and its `__class__` attributes are used to create or update a mapping in the registry. The implication of this is for any `__rdf_resource__`, there can only be one blue print (`__class__`) representing it. The consequence of this is parsing and validating dataclasses follow the SOLID principles. Serialised instances of a child schema should be serialisable and deserialisable to the parent class. Common fields shall have their serialised values updated based on `FieldInfo`, and fields that are present in one but not in another will be kept as is. 

This means as long as users extend the basic miappe classes, the serialised RO-Crate should be miappe compatible. 

We will demonstrate this with an example:

```python
ChildInvestigation = create_linked_dataclass(schema=new_schema, base = Investigation) # fake mthd
child_instance = ChildInvestigation(....)

# Serialise to dict 
child_obj: dict[str, Any] = child_instance.encode()

# Create parent instance 
parent_instance = Investigation.decode(obj=child_obj, schema= ChildInvestigation.__schema__)
parent_obj = Investigation.encode(parent_obj)

# Convert back to child instance 
deserialised_child_instance = ChildInvestigation.decode(obj=parent_obj, schema=Investigation.__schema__)

# Should work 
assert child_instance == deserialised_child_instance
``` 
I cannot think of a scenario where users will have two subclasses `InvestigationA` and `InvestigationB` for a single project that warrants a non-singleton  registry for mapping. In the unlikely event that this happens, they can create `InvestigationC` that has all fields from both `InvestigationA` and `InvestigationB`. 

## Session

Session represents an open connection to a store. 

```python
registry = Registry() # Singleton

# Create a session to my_project_ro_crate.json
session = registry.session("my_project_ro_crate.json")

# Create data 
experiment = Experiment(...)
plants = [Plant(id=f"Plant_{i}", variety="barley_alpha", experiment=experiment) for i in range(10)]
experiment.plants = [plant.id for plant in plants]
# Add to session 
session.add(experiment) # Should automatically add all referenced plants 
# Serialise 
session.commit() 
```

## Session ID Rules 
- Within a session, two instances A and B can have the same ID iff A and B are separate resources that are not subclasses of one another - i.e. `not issubclass(A.__class__, B.__class__) and not issubclass(B.__class__, A.__class__)` (cover scenario 4 of #3)
- If two instances with the same ID whose `.__class__` attrs having a parent-child relation are added to the session - i.e. similar to scenarios 1, 2, 3 of issue #3, the behaviour can vary depending on a session config: 
  - `on_clashing_id = "error"`, will raise an error
  - `on_clashing_id = "warning"`, will raise a warning and overwrite the instance based on recency 
  - `on_clashing_id = "ignore"`, will do nothing and overwrite the instance based on recency
